src/core.cpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * Mozart++ Template Library
+ * Licensed under MIT License
+ * Copyright (c) 2020 Covariant Institute
+ * Website: https://covariant.cn/
+ * Github:  https://github.com/covariant-institute/
+ */
+
+#include <mozart++/core>
+
+#ifdef MOZART_PLATFORM_WIN32
+
+#include <shlobj.h>
+
+#pragma comment(lib, "shell32.lib")
+
+#endif
+
+#ifdef _MSC_VER
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+
+#ifndef _WIN32_WINNT
+#define _WIN32_WINNT 0x0501
+#endif
+
+#include <windows.h>
+#include <Dbghelp.h>
+#pragma comment(lib, "DbgHelp")
+
+namespace mpp {
+    std::string cxx_demangle(const char* name)
+    {
+        char buffer[1024];
+        DWORD length = UnDecorateSymbolName(name, buffer, sizeof(buffer), 0);
+        if (length > 0)
+            return std::string(buffer, length);
+        else
+            return name;
+    }
+}
+
+#elif defined __GNUC__
+
+#include <cxxabi.h>
+
+namespace mpp {
+    std::string cxx_demangle(const char *mangled) {
+        char buffer[1024] = {0};
+        size_t size = sizeof(buffer);
+        int status;
+        char *ret = abi::__cxa_demangle(mangled, buffer, &size, &status);
+        if (ret != nullptr)
```

Comment:
Please do not use `char buffer[1024]` in case that the typename length are greater than 1023
(it's possible as we have `mpp::typelist` which can consist of arbitaty count types).

and it's not safe because the [GNU ABI](https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html) documentation said:
> char * __cxa_demangle (const char *mangled_name, char *output_buffer, size_t *length, int *status);
>
> If output_buffer is not long enough, it is expanded using realloc.
>

so the unsafe point is that out output_buffer are not allocated with `malloc`, let alone `realloc`.

My suggesstion is that:
```cpp
char *ptr = abi::__cxa_demangle(name, nullptr, nullptr, nullptr);
if (ptr) {
    auto s = std::string(ptr);
    std::free(ptr);
    return std::move(s);
}
```

Let the implementation to detect how many space are needed to hold the whole typename.
as the documentation also said:
> output_buffer may instead be NULL; in that case, the demangled name is placed in a region of memory allocated with malloc.

Finally, I don't know what's the implenmentation of Windows API, be aware.